Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp

```diff
@@ -226,25 +226,26 @@ namespace Terrain
             auto& surfaceDataContext = SurfaceData::SurfaceDataSystemRequestBus::GetOrCreateContext(false);
             typename SurfaceData::SurfaceDataSystemRequestBus::Context::DispatchLockGuard scopeLock(surfaceDataContext.m_contextMutex);
 
-            for (int32_t y = yStart; y < yEnd; y++)
+            auto perPositionCallback = [this, &pixels]
+                ([[maybe_unused]] size_t xIndex, [[maybe_unused]] size_t yIndex,
+                const AzFramework::SurfaceData::SurfacePoint& surfacePoint,
+                [[maybe_unused]] bool terrainExists)
             {
-                for (int32_t x = xStart; x < xEnd; x++)
-                {
-                    bool terrainExists = true;
-                    float terrainHeight = 0.0f;
-                    float xPos = x * m_sampleSpacing;
-                    float yPos = y * m_sampleSpacing;
-                    AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
-                        terrainHeight, &AzFramework::Terrain::TerrainDataRequests::GetHeightFromFloats,
-                        xPos, yPos, AzFramework::Terrain::TerrainDataRequests::Sampler::EXACT, &terrainExists);
-
-                    const float clampedHeight = AZ::GetClamp((terrainHeight - m_terrainBounds.GetMin().GetZ()) / m_terrainBounds.GetExtents().GetZ(), 0.0f, 1.0f);
-                    const float expandedHeight = AZStd::roundf(clampedHeight * AZStd::numeric_limits<uint16_t>::max());
-                    const uint16_t uint16Height = aznumeric_cast<uint16_t>(expandedHeight);
-
-                    pixels.push_back(uint16Height);
-                }
-            }
+                const float clampedHeight = AZ::GetClamp((surfacePoint.m_position.GetZ() - m_terrainBounds.GetMin().GetZ()) / m_terrainBounds.GetExtents().GetZ(), 0.0f, 1.0f);
+                const float expandedHeight = AZStd::roundf(clampedHeight * AZStd::numeric_limits<uint16_t>::max());
+                const uint16_t uint16Height = aznumeric_cast<uint16_t>(expandedHeight);
+
+                pixels.push_back(uint16Height);
+            };
+
+            AZ::Vector2 stepSize(m_sampleSpacing);
+            AZ::Vector3 maxBound(
+                m_dirtyRegion.GetMax().GetX() + m_sampleSpacing, m_dirtyRegion.GetMax().GetY() + m_sampleSpacing, 0.0f);
+            AZ::Aabb region;
+            region.Set(m_dirtyRegion.GetMin(), maxBound);
```

Comment:
In this case, the loop goes from `m_dirtyRegion.GetMin().Get<X/Y>() / m_sampleSpacing` to `(m_dirtyRegion.GetMax().Get<X/Y>() / m_sampleSpacing) + 1`. In the `ProcessRegion` set of functions, we use `boundingBox.GetExtents().Get<X/Y>() / m_sampleSpacing`. This works out to one less sample in the ProcessRegion functions. In order to keep the existing behavior the same, I add one additional sample spacing here to the max bound. Should the ProcessRegion functions process one additional sample in all cases? Or is this a special case for the terrain feature processor?

Reply:
I'm not 100% sure, but this probably has to do with whether or not ProcessRegion() should be inclusive on both sides [min, max] or just inclusive on one side [min, max).  I _think_ we want the latter so that when processing adjacent regions we don't process the same edge twice, but we might need to revisit this at some point to ensure we've made a deliberate choice that makes sense with our use cases.  (We'll also _definitely_ want to document it either way)  

So, I _think_ you're doing the right thing here by special-casing the extra iteration just for the Feature Processor, where it probably needs that boundary pixel for correct blending or something.

Reply:
> What if we added the ability to request the width / height / total samples for a region based call before actually calling it? Then you could reserve all the memory you're going to need first.

Instead of having a separate call to request what the number of samples, etc., will be, how about adding another optional callback to the `ProcessRegion` functions? So, the `ProcessHeightsFromRegion` function would have the following signature:
```
void ProcessHeightsFromRegion(const AZ::Aabb& inRegion,
            const AZ::Vector2& stepSize,
            AzFramework::Terrain::SurfacePointRegionFillCallback perPositionCallback,
            Sampler sampleFilter = Sampler::DEFAULT,
            AzFramework::Terrain::NumSamplesCallback numSamplesCallback)
```
Before the ProcessRegion function starts processing individual positions, it will first call the `numSamplesCallback` with the number of X/Y samples which can be used to allocate memory.

Reply:
I think a separate API would be easier to read and use.  Assuming the common the use case for the info would be:
* get num samples
* allocate a structure to store results
* Process from region, filling in the structure with results

IMO it would be easier to read the above flow linearly in the code, rather than having the Process call run two separate steps at different times, where one callback is called once, and one is called N times.

Reply:
I'm pretty fine with either approach but I think I agree with @mbalfour-amzn that just having a separate API would be cleaner.